### DISCUSSION TITLE
[Feature request] Support multiple elasticsearch nodes

### DISCUSSION THREAD

**Discussion by dploeger** [grafana/grafana#58399] [Data Sources]
Grafana version: 4.1.1
Datasource: Elasticsearch 2.3.5
OS: macOS 10.12.4

The elasticsearch data source currently only seems to support a single ES-node. Usually, elasticsearch comes as a cluster with multiple nodes. It would be great, if one could specify multiple nodes, either in a master/fallback a real round-robin way.

The other possibility would be to create a high available load balancer in front of them, but it would be much nicer and easier, if the application supports it out of the box.

**Comment by torkelo** [comment]
usually, you have gateway nodes that act like load balancers to the ES cluster. This is how I and other usually setup ES with Grafana or Kibana. 

Is this not the norm anymore?

**Comment by dploeger** [comment]
I specifically want to integrate Grafana into a Graylog setup. Graylog uses the zen unicast discovery to hook up to the ES cluster and I don't need another load balancer.

**Comment by frittentheke** [comment]
Service discovery in contrast to the current static config is one other thing, but at least having the ability to configure a list of ES nodes to provide some redundancy in hooking up Grafana to an ES cluster would be nice. Maybe even a little load balancing could be done by doing queries round robin, just like suggested.

Also auto-filling the list by querying for all ES nodes with just a single node configures, just like logstash does it (https://www.elastic.co/guide/en/logstash/current/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-sniffing) would be a nice addition.

**Comment by lpic10** [comment]
I think the feature request is about adding "sniffing" as it exists on ruby/golang/java/javascript libraries for Elasticsearch, in addition to allowing setting more than one node in the main configuration. 

https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/configuration.html
https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/transport-client.html
https://github.com/elastic/elasticsearch-ruby/tree/master/elasticsearch-transport
https://github.com/olivere/elastic/wiki/Sniffing

**Comment by dploeger** [comment]
The original request was just about adding more than one nodes, but sniffing should be the way to go, I think.

**Comment by mmolinac** [comment]
I have the same opinion (and need) that @dploeger .
I think it would be a good feature to ask a list of servers without needing a load balancer.

**Comment by sureshkumar1984** [comment]
+1

**Comment by yossiv** [comment]
Hi guys, can we bring live this feature?
this would be great instead to use loadbalancer 
like this config example that was taken from Graylog conf file:
 #elasticsearch_hosts = http://10.120.55.81:9200,http://10.120.55.82:9200,http://10.120.55.83:9200
since if the master is down for some reason, it needs to go to the next member in the cluster. otherwise we have single point of failure

**Comment by alxshr** [comment]
+1

**Comment by shizacat** [comment]
+1

**Comment by magaton** [comment]
+1

**Comment by ManinderHSC** [comment]
+1

**Comment by buroa** [comment]
I implemented this by deploying a HAProxy within my Grafana nodes, and using that to load balance between ELBs located in multiple regions. Grafana then points to its local HAProxy as a Elasticsearch data source.

**Comment by milandamen** [comment]
@buroa That adds another system that can fail, and it adds complexity and maintenance. It would be much better for Grafana to do the load balancing instead.

**Comment by frittentheke** [comment]
Without having looked at the ES client code Grafana uses ... the clients, i.e. for Python which Elastic provides also contain the sniffing / connection pool logic with all the details required to use and apply the cluster state - https://github.com/elastic/elasticsearch-php/blob/master/src/Elasticsearch/ConnectionPool/SniffingConnectionPool.php
Elastic is doing a GO client, but that currently lacks this very feature https://github.com/elastic/go-elasticsearch/blob/master/client/option.go#L93

This does not make having a list of ES hosts as @yossiv said, obsolete. Certainly there needs to be redundancy for the initial connection, but after that the sniffing should take over. Maybe it's even sensible to save / cache the last list from the sniffing to resume on the next connection to ES. This would allow for an even greater redundancy once a first successful connection to any node in the ES cluster was made.

**Comment by buroa** [comment]
> @buroa That adds another system that can fail, and it adds complexity and maintenance. It would be much better for Grafana to do the load balancing instead.

Take a look at  the diagram and then reply.

![picture1](https://user-images.githubusercontent.com/36205263/46617830-c2da1e80-caeb-11e8-95bf-e0d0fa6e69f5.png)

**I did add another failover channel for MySQL so Grafana can point to it's local HAProxy which routes to either region -- which is not shown here.**

**Comment by yossiv** [comment]
Hi if someone has a free F5 LTM or even a free open source loadblancer .
here is how i did it.
i put a VIP that represend the elastic node query pool members, 
and on the grafana im putting the Virtual ip of the LB, so the Grafana calling to the VIP, and the Loadbalancer hundling the calls. 
![image](https://user-images.githubusercontent.com/19520218/46630095-cef5c880-cb4b-11e8-9c51-cb38bf705f93.png)

**Comment by hrak** [comment]
All the "i solved this using load balancing solution X" replies here are super interesting, but add nothing to the initial feature request of supporting multiple Elasticsearch client nodes (and/or sniffing). Practically all ES client libraries (python/ruby/golang) support this.

**Comment by beingyash** [comment]
Has anyone figured out if there's some challenge involved in the implementation of this feature, or the grafana people don't find it a use-case enough to be bothered about it or do they not deem it important enough to be prioritized yet?

**Comment by marefr** [comment]
Not enough upvotes/interest shown for us to put it on our roadmap and work on it. There's probably challenges as well. Contributions are welcome if someone is interested. Let us know if you need/want some pointers to the code.

**Comment by frittentheke** [comment]
@marefr  looking at the code for DS of type "elasticsearch" (https://github.com/grafana/grafana/blob/39d8e73412fc7b065d3ece76c143699697666aa4/pkg/tsdb/elasticsearch/client/client.go#L30), it appears you are NOT using the Golang lib from Elastic.co  but rather use a "regular" HTTP client.

This is unfortunate, because the "official" client supports node discovery (as suggested in this very issue here), right out of the box: https://github.com/elastic/go-elasticsearch/blob/master/estransport/estransport.go#L64

Apart from that, a lot of things regarding robustness of the connection handling or features such aas connection pooling are all already supported by the golang client.

**Comment by ominolab** [comment]
+1

**Comment by cabrizio** [comment]
so it's enough to put an Nginx on top of my 3 elasticsearch nodes to fix this right?

**Comment by milandamen** [comment]
Any solution suggesting a single load balancer to balance to the Elasticsearch nodes is missing the point of this issue. If you have a single load balancer, that is still a single point of failure, and you could just have connected to a single Elasticsearch node instead.

The only way to properly support redundancy in a way that would eliminate a single point of failure would be to let Grafana connect to multiple Elasticsearch nodes.

**Comment by davkal** [comment]
Let me understand this correctly: do we want a list of nodes for failover (asking the next node if the first one does not reply) or do we want round-robin to balance load? Or both?

**Comment by frittentheke** [comment]
@davkal this is not just about doing a list and fail over to the next host.
Elasticsearch clients (or their library for that matter) explicitly support having multiple ElasticSearch nodes and also use those to do "no loadbalancer"-balancing of requests, see my comment at https://github.com/grafana/grafana/issues/8170#issuecomment-566136959

This is quite in line with other "databases" which have smart clients to (smartly) balance requests to multiple nodes of a cluster - i.e. Cassandra https://docs.datastax.com/en/dse-planning/doc/planning/planningAntiPatterns.html#planningAntiPatterns__AntiPatLoadBal or rather its client libraries https://docs.datastax.com/en/developer/java-driver/4.8/manual/core/load_balancing/

**Comment by hackery** [comment]
Is this a special case of "allow configuring multiple back-end servers for datasources"? I have had the same thought about Influx (we pair them for redundancy, duplicating metrics into each) and #3676 is a similar request for mysql/postgres, so there's some commonality.

The main "pro" argument I think is to avoid a separate load balancer - a monitoring system needs to have as few dependencies as possible, as it's probably reporting on those same network elements, and relying on something you're reporting on in order to report creates fragility.

I suppose the main "con" argument is "doing load balancing properly is hard - that's why people build businesses from selling these things" - but how much would be needed for a useful feature? Failover management would be relatively simple; round-robin a bit more work; load-based balancing harder; for my use case even failover capability would be very useful.

Elasticsearch is perhaps a bit unique in the Grafana context, because of the "sniffing" capability, but it would still be useful to include multiple coordinating nodes as starting points. The UI for maintaining a list of servers doesn't need to expose that additional hosts are then discovered.

**Comment by milandamen** [comment]
@davkal FYI.

I agree with @hackery . I think this is a generally wanted feature for multiple kinds of datasources. Making a generic fail-over or round-robin system would probably benefit all cluster-based or replication-based datasources, though coupling that with the official golang elasticsearch client would probably be best.

I think this feature request can be summed up as:

"We want Grafana to be able to send request to multiple Elasticsearch nodes, because currently when the node that Grafana uses goes down, you won't be able to query any data from the cluster and all graphs go blank."

A fail-over to a next node would be a basic solution to this problem.
A round-robin system would be nice to have but wouldn't be required.

If just a fail-over is implemented, a round-robin could be implemented later, and in my opinion balancing the load to the nodes is less important than a fail-over, because Grafana already only queries a single node.

**Comment by Mandark-droid** [comment]
+1 it would be a good feature to have and would really help distribute the load across node for large clusters. And if you implement it using the elastic.go client we can also use the split function to distribute the load for large clusters. I have multiple such clusters with total size of 50TB+ indexes.

**Comment by gelicia** [comment]
Hello! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

This discussion was converted into an issue, which is available [here](https://github.com/grafana/grafana/issues/73449).
